core/modules/locale/locale.module

```diff
@@ -563,12 +563,21 @@ function locale($string = NULL, $context = NULL, $langcode = NULL) {
         // Refresh database stored cache of translations for given language.
         // We only store short strings used in current version, to improve
         // performance and consume less memory.
-        $result = db_query("SELECT s.source, s.context, t.translation, t.language FROM {locales_source} s LEFT JOIN {locales_target} t ON s.lid = t.lid AND t.language = :language WHERE s.version = :version AND LENGTH(s.source) < :length", array(':language' => $langcode, ':version' => BACKDROP_VERSION, ':length' => config_get('locale.settings', 'cache_length')));
-        foreach ($result as $data) {
-          $locale_t[$langcode][$data->context][$data->source] = (empty($data->translation) ? TRUE : $data->translation);
+        try {
+          $result = db_query("SELECT s.source, s.context, t.translation, t.language FROM {locales_source} s LEFT JOIN {locales_target} t ON s.lid = t.lid AND t.language = :language WHERE s.version = :version AND LENGTH(s.source) < :length", array(
+            ':language' => $langcode,
+            ':version' => BACKDROP_VERSION,
+            ':length' => config_get('locale.settings', 'cache_length'),
+          ));
+          foreach ($result as $data) {
+            $locale_t[$langcode][$data->context][$data->source] = (empty($data->translation) ? TRUE : $data->translation);
+          }
+          cache()->set('locale:' . $langcode, $locale_t[$langcode]);
+          lock_release('locale_cache_' . $langcode);
+        }
+        catch (PDOException $e) {
+          return $string;
```

Comment:
We are simply returning the string passed to this function here, which may also be `NULL`. Shouldn't there be a meaningful error returned here instead?

Reply:
The string is correct - it's untranslated if something weird happened in the query. (The function docblock explains a bit more about the meaning of NULL here.)

If you expand below code a bit more, you'll find the same logic again in existing code.